Iox #1142 icediscovery in c (#1165)
eclipse-iceoryx/iceoryx: iceoryx_examples/icediscovery_in_c/README.md

services and one searching for these making different search queries. The
behavior and structure is quite similar to the [icediscovery C++ example](https://github.com/eclipse-iceoryx/iceoryx/tree/v2.0.0/iceoryx_examples/icedelivery).

<!--## Expected Output-->
<!-- @todo Add expected output with asciinema recording before v2.0-->

## Code walkthrough

### Offer services

We create several publisher ports which offer their services on construction by
default. For more dynamism the `cameraPublishers` offer/stop their services
periodically. If you want more information on how to create publisher ports,
have a look at the [icedelivery C example](https://github.com/eclipse-iceoryx/iceoryx/tree/v2.0.0/iceoryx_examples/icedelivery_in_c).

### Find services

To be able to search for services, we have to create a service discovery handle
which points to `storage`, the memory location where the service discovery is
stored.

<!--[geoffrey][iceoryx_examples/icediscovery_in_c/iox_c_find_service.c][create service discovery handle]-->
```c
iox_service_discovery_storage_t storage;
iox_service_discovery_t serviceDiscovery = iox_service_discovery_init(&storage);
```

> It is included via:

Reviewer: What is it? The service discovery? The handle? ... 

I would somehow start with this (to use service discovery, we need to include ...) before mpving on to service discovery creation/initialization.

<!--[geoffrey][iceoryx_examples/icediscovery_in_c/iox_c_find_service.c][include service discovery]-->
```c
#include "iceoryx_binding_c/service_discovery.h"
```

We can now call three different find service functions. Let's start with

<!--[geoffrey][iceoryx_examples/icediscovery_in_c/iox_c_find_service.c][find service and apply callable]-->
```c

> iox_service_discovery_find_service_apply_callable(

Reviewer: The terminology of callable is more of a C++ one, in C only functions and function pointers are callable and we can only pass the latter (there are no C++ functors). Still, I would keep the name as it is correct but maybe unusual to a C user.

    serviceDiscovery, "Radar", "FrontLeft", "Image", printSearchResult, MessagingPattern_PUB_SUB);
```

which searches for all `{Radar, FrontLeft, Image}` services offered by
publishers and applies the provided function on each of them. The function must
have the signature `void(const iox_service_description_t)`. Here we pass a
function that prints the found services on the console:

<!--[geoffrey][iceoryx_examples/icediscovery_in_c/iox_c_find_service.c][print function to be applied to search results]-->
```c
void printSearchResult(const iox_service_description_t service)
{
    printf(
        "- Service: %s, Instance: %s, Event: %s\n", service.serviceString, service.instanceString, service.eventString);
}
```

> We can not only search for exactly matching services, but also for wildcards

Reviewer 2: , in combination with wildcards (`NULL`) we can search also for all instances and events with the service `Radar`.

> We can not only search for exactly matching services, but also for wildcards

Reviewer 2: We can not only search for exact matching services,

using `NULL`:

<!--[geoffrey][iceoryx_examples/icediscovery_in_c/iox_c_find_service.c][search for all Radar services]-->
```c
iox_service_discovery_find_service_apply_callable(
    serviceDiscovery, "Radar", NULL, NULL, printSearchResult, MessagingPattern_PUB_SUB);

> ```

Reviewer 2: The wildcard can be used for all strings which describe a service, e.g. service, instance and event.

With the above call we look for every `Radar` service with any instance and any
event.

Let's now search for all `Camera` services and store the results in an array:

<!--[geoffrey][iceoryx_examples/icediscovery_in_c/iox_c_find_service.c][search for all Camera services]-->
```c
numberFoundServices = iox_service_discovery_find_service(serviceDiscovery,
                                                         "Camera",
                                                         NULL,
                                                         NULL,
                                                         searchResult,
                                                         SEARCH_RESULT_CAPACITY,
                                                         &missedServices,
                                                         MessagingPattern_PUB_SUB);
```

`searchResult` is a `iox_service_description_t` array of size

> `SEARCH_RESULT_CAPACITY`. The found services are written into this array, but

Reviewer 2: The matching services are written

> `SEARCH_RESULT_CAPACITY`. The found services are written into this array, but

Reviewer: ```suggestion
`SEARCH_RESULT_CAPACITY`. The found services are written into this array
```

> at maximum `SEARCH_RESULT_CAPACITY`. If the number of found services exceeds

Reviewer: ```suggestion
up to a maximum of `SEARCH_RESULT_CAPACITY`. If the number of found services exceeds
```

> the array's capacity, the number of not stored services is written to

Reviewer: ```suggestion
the array's capacity, the number of services that could not be stored is written to
```

`missedServices`. The number of stored services is returned. Since the
`cameraPublishers` periodically offer/stop their services, you should see
sometimes 5 `Camera` services and sometimes none.

Finally, let's try out the third find service function. We search again for all
`Camera` services but additionally count the front camera services:

<!--[geoffrey][iceoryx_examples/icediscovery_in_c/iox_c_find_service.c][search for all front camera services]-->
```c
iox_service_discovery_find_service_apply_callable_with_context_data(
    serviceDiscovery, "Camera", NULL, NULL, searchFrontDevices, &numberFrontCameras, MessagingPattern_PUB_SUB);
```

This function is quite similar to the first find service function, but we

> pass an additional argument:

Reviewer 2: pass an additional argument `numberFrontCameras`.

<!--[geoffrey][iceoryx_examples/icediscovery_in_c/iox_c_find_service.c][store number of front cameras]-->
```c
uint32_t numberFrontCameras = 0U;
```

We use this variable to store the number of front cameras and provide it as

> second parameter to the function which is applied to all found services:

Reviewer 2: and provide a pointer to it as second argument `count` to the function `searchFrontDevices` which is applied

> second parameter to the function which is applied to all found services:

Reviewer: Maybe it is more important that it is an output parameter to return the count. Do we want to add this information or is this clear from usage? To me it is but I may be biased.

Author: I find this clear from usage. Additionally we state `We use this variable to store the number...`

instance and event.

Let's now search for all `Camera` services and store the results in an array:

<!--[geoffrey][iceoryx_examples/icediscovery_in_c/iox_c_find_service.c][search for all Camera services]-->
```c
numberFoundServices = iox_service_discovery_find_service(serviceDiscovery,
                                                         "Camera",
                                                         NULL,
                                                         NULL,
                                                         searchResult,
                                                         SEARCH_RESULT_CAPACITY,
                                                         &missedServices,
                                                         MessagingPattern_PUB_SUB);
```

`searchResult` is a `iox_service_description_t` array of size
`SEARCH_RESULT_CAPACITY`. The matching services are written into this array, up
to a maximum of `SEARCH_RESULT_CAPACITY`. If the number of found services
exceeds the array's capacity, the number of services that could not be stored is
written to `missedServices`. The number of stored services is returned. Since

> the `cameraPublishers` periodically offer/stop their services, you should see

Reviewer: ```suggestion
the `cameraPublishers` periodically offer/stop their services, we should see
```

service discovery. It provides two applications - one offering different
services and one searching for those with different search queries. The
behavior and structure is quite similar to the [icediscovery C++ example](https://github.com/eclipse-iceoryx/iceoryx/tree/v2.0.0/iceoryx_examples/icedelivery).

<!--## Expected Output-->
<!-- @todo Add expected output with asciinema recording before v2.0-->

## Code walkthrough

### Offer services

We create several publisher ports which offer their services on construction by
default. For more dynamism the `cameraPublishers` offer/stop their services
periodically. If you want more information on how to create publisher ports,
have a look at the [icedelivery C example](https://github.com/eclipse-iceoryx/iceoryx/tree/v2.0.0/iceoryx_examples/icedelivery_in_c).

### Find services

To be able to search for services, we need to include:

<!--[geoffrey][iceoryx_examples/icediscovery_in_c/iox_c_find_service.c][include service discovery]-->
```c
#include "iceoryx_binding_c/service_discovery.h"
```

To get a handle to the service discovery we call `iox_service_discovery_init`.
The service discovery requires memory of the amount of

> `sizeof(iox_service_discovery_storage_t)` to be initialized.

Reviewer: Ok, but this is hidden from the user and an implementation detail. From my point of view the important information is that the user must provide an instance of `iox_service_discovery_storage_t` to an initialization function. The size of objects of this class is not relevant to the user and taking care of by the struct itself.

Technically the user can pass any buffer of sufficient size (requires a cast), but it is not intended to be used that way...

I would hence not mention the detail wit `sizeof`, but not a must change ...

> `sizeof(iox_service_discovery_storage_t)` to be initialized.

Reviewer: Technically the only information relevant to the user is that she has to pass a struct of type `iox_service_discovery_storage_t` which will be of the correct size (by construction/API). 

Hence one could leave out the information with `sizeof`. It may give people ideas to use it with raw buffers of that size (which will work if correctly aligned but is not encouraged).

Not a must change but should probably be omitted.
